### ISSUE TITLE
`@traitfn` on callable types

### ISSUE THREAD

**Issue by MilesCranmer** [mauro3/SimpleTraits.jl#86]
There is a workaround for this but in case you'd like to support it, the following syntax does not work. My use-case is to build a generic constructor for any type `D` satisfying some trait:

```julia
@traitfn (::Type{D})(args...) where {D; IsMyTrait{D}} = f(args...)
```

This will raise the following error:

```julia
ERROR: syntax: invalid "::" syntax around /Users/mcranmer/.julia/packages/SimpleTraits/WyNKl/src/SimpleTraits.jl:338
Stacktrace:
 [1] top-level scope
   @ REPL[27]:1
```

The workaround is to pass a symbol:

```julia
@traitfn (d::Type{D})(args...) where {D; IsMyTrait{D}} = f(args...)
```
which works.

**Comment by MilesCranmer** [comment]
As in https://github.com/mauro3/SimpleTraits.jl/issues/87, maybe I should just avoid doing stuff like this, since it will invalidate functions in Base.

**Comment by mauro3** [comment]
Yes, this is also a bit iffy.  Again, similar reasoning to #87 should mean that it should work in principle (i.e. don't forget to define the `Not` version, preferably before the not-`Not` version).  But again, I may not have thought this through.

Regarding the syntax error: it would be ok to fix this as it could be use for more benign applications too.  PR welcome.

**Comment by mauro3** [comment]
I'm not following here:

> But unless there's a way to modify multiple dispatch itself (throughout Julia) to check the parametric type of AbstractTraitWrapper, it doesn't seem possible...

You don't refer to above example with that, right?

**Comment by MilesCranmer** [comment]
Sorry I was thinking of the example in #87, not the one in here.
